File: Data structure/Linked List/singly_linked_list/js/linklist.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Linklist {
  constructor() {
    this.head = null;
    this.count = 0;
  }

  //MEHTHOD TO INSERT AT START
  insert_at_start(data) {
    let new_node = new Node(data);
    this.count++;

    if (this.head === null) {
      this.head = new_node;
      return;
    }

    new_node.next = this.head;
    this.head = new_node;
  }

  //MEHTHOD TO INSERT AT END
  insert_at_end(data) {
    let new_node = new Node(data);
    this.count++;

    if (this.head === null) {
      this.head = new_node;
      return;
    }

    let temp = this.head;
    while (temp.next != null) temp = temp.next;

    temp.next = new_node;
  }

  //MEHTHOD TO INSERT AT ANY POSITION
  insert_at_position(position, data) {
    if (position < 1 || position > this.count + 1) {
      console.log("Invalid position");
      return;
    }

    if (position === this.count + 1) {
      this.insert_at_end(data);
      return;
    }
    if (position === 1) {
      this.insert_at_start(data);
      return;
    }
    this.count++;
    let temp = this.head;
    let new_node = new Node(data);
    let index = 1;
    position -= 1;
    while (index != position) {
      temp = temp.next;
      index++;
    }

    new_node.next = temp.next;
    temp.next = new_node;
  }

  //MEHTHOD TO DELETE FROM BEGINNING
  delete_from_beginning() {
    if (this.count === 0) {
      console.log("Empty list");
      return;
    }
    this.count--;
    this.head = this.head.next;
  }

  //MEHTHOD TO DELETE FROM END
  delete_from_end() {
    if (this.count === 0) {
      console.log("Empty list");
      return;
    }
    this.count--;
    if (this.count === 1) {
      this.delete_from_beginning();
      return;
    }

    let temp = this.head;
    while (temp.next.next != null) temp = temp.next;
    temp.next = null;
  }

  //MEHTHOD TO DELETE FROM ANY GIVEN POSITION
  delete_from_position(position) {
    if (position < 1 || position > this.count) {
      console.log("Invalid position");
      return;
    }

    if (position === this.count) {
      this.delete_from_end();
      return;
    }
    if (position === 1) {
      this.deelte_from_beginning();
      return;
    }

    this.count--;
    let temp = this.head;
    position -= 1;
    let curr_index = 1;

    while (curr_index != position) {
      temp = temp.next;
      curr_index++;
    }

    temp.next = temp.next.next;
  }

  //METHOD RETRIVES THE DATA FROM A GIVEN POSITION
  get_data_from(position) {
    if (position < 1 || position > this.count) {
      console.log("Invalid Position.");
      return;
    }

    let index = 1;
    let ptr = this.head;
    while (index != position) {
      index++;
      ptr = ptr.next;
    }

    return ptr.data;
  }

  //METHOD TO PRINT THE LIST
  static print_list(list) {
    let temp = list.head;
    while (temp) {
      process.stdout.write(temp.data + "->");
      temp = temp.next;
    }
    console.log("null");
  }

  //METHOD RETURNS THE SIZE OF LIST
  list_size() {
    return this.count;
  }

  // METHOD TO CLEAR THE LIST
  static clear_list(list) {
    list.head = null;
    list.count = 0;
  }
}

module.exports = {
  Linklist,
};

File: Data structure/Linked List/singly_linked_list/python/linklist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Linklist:
    def __init__(self):
        self.head = None
        self.count = 0

    # METHOD TO INSERT AT THE START
    def insert_at_beginning(self, data):
        new_node = Node(data)
        self.count += 1

        if(self.head == None):
            self.head = new_node
            return

        new_node.next = self.head
        self.head = new_node

    # METHOD TO INSERT AT THE END
    def insert_at_end(self, data):
        new_node = Node(data)
        self.count += 1

        if(self.head == None):
            self.head = new_node
            return

        ptr = self.head

        while(ptr.next):
            ptr = ptr.next

        ptr.next = new_node

    # METHIOD TO INSERT AT ANY ARBITARY POSITION
    def insert_at_position(self, position, data):
        if(position < 1 or position > self.count+1):
            print("Invalid position.")
            return

        if(position == 1):
            self.insert_at_beginning(data)
            return
        if(position == self.count+1):
            self.insert_at_end(data)
            return
        self.count += 1
        new_node = Node(data)
        ptr = self.head
        index = 1
        position -= 1
        while(index != position):
            index += 1
            ptr = ptr.next

        new_node.next = ptr.next
        ptr.next = new_node

    # METHOD TO DELETE FROM BEGINNING
    def delete_from_start(self):
        if(self.count == 0):
            print("The list is empty.")
            return

        self.count -= 1
        self.head = self.head.next

    # METHOD TO DELETE FROM THE END
    def delete_from_end(self):
        if(self.count == 0):
            print("The list is empty.")
            return

        self.count -= 1
        if(self.count == 1):
            self.head = None
            return

        ptr = self.head
        while ptr.next.next:
            ptr = ptr.next
        ptr.next = None

    # METHOD TO DELETE FROM ANY ARBITARY POSITION
    def delete_from_position(self, position):
        if(position < 1 or position > self.count):
            print("Invalid position.")
            return

        if(position == 1):
            self.delete_from_start()
            return
        if(position == self.count):
            self.delete_from_end()
            return

        self.count -= 1
        ptr = self.head
        index = 1
        position -= 1
        while index != position:
            index += 1
            ptr = ptr.next
        ptr.next -= ptr.next.next

    # METHOD TO RETRIEVE DATA FROM POSITION
    def get_data_from(self, position):
        if(position < 1 or position > self.count):
            print("Invalid position.")
            return None

        ptr = self.head
        index = 1
        while(index != position):
            index += 1
            ptr = ptr.next

        return ptr.data
    # METHOD TO PRINT THE LIST

    @staticmethod
    def print_list(list):
        temp = list.head
        while(temp):
            print(temp.data, " --> ", end="")
            temp = temp.next
        print("NULL")

    # METHOD RETURS THE NUMBER OF ELEMETS IN THE LIST
    def size(self):
        return self.count

    # METHOD TO CLAR THE LIST
    @staticmethod
    def clear(list):
        list.head = None
        list.count = 0
